crates/lib/src/keyboard.rs

```diff
@@ -0,0 +1,125 @@
+//! Event source for keyboard input and related events
+use async_priority_channel as priority;
+use tokio::{
+	io::AsyncReadExt,
+	sync::{mpsc, oneshot, watch},
+};
+use tracing::trace;
+
+use crate::{
+	error::{CriticalError, RuntimeError},
+	event::{Event, Priority, Source, Tag},
+};
+
+/// The configuration of the [keyboard][self] worker.
+///
+/// This is marked non-exhaustive so new configuration can be added without breaking.
+#[derive(Debug, Clone, Default)]
+#[non_exhaustive]
+pub struct WorkingData {
+	/// Whether or not to watch for 'end of file' on stdin
+	pub eof: bool,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+/// Enumeration of different keyboard events
+pub enum Keyboard {
+	/// Event representing an 'end of file' on stdin
+	Eof,
+}
+
+/// Launch the filesystem event worker.
+///
+/// While you can run several, you should only have one.
+///
+/// Sends keyboard events via to the provided 'events' channel
+pub async fn worker(
+	mut working: watch::Receiver<WorkingData>,
+	errors: mpsc::Sender<RuntimeError>,
+	events: priority::Sender<Event, Priority>,
+) -> Result<(), CriticalError> {
+	let mut send_close = None;
+	while working.changed().await.is_ok() {
+		let watch_for_eof = { working.borrow().eof };
+		match (watch_for_eof, send_close.take()) {
```

Comment:
Hmm, maybe the refactor to a match makes this 'take' problematic? Might need to go back.

Reply:
I've refactored to do the 'take' in an inner if-let. Not sure if that is necessary but I think it is.

Please make any changes that you feel are necessary and merge if you get to a happy state. Though the question about the Keyboard error is still outstanding for the moment.